File: javascript/dnd-character/dnd-character.js
export const abilityModifier = n =>
    n < 3 ? (() => { throw new Error('Ability scores must be at least 3') })() :
    n > 18 ? (() => { throw new Error('Ability scores can be at most 18'); })() :
    Math.floor((n - 10) / 2);


export class Character {
    characteristics = [
        'strength', 'dexterity', 'constitution',
        'intelligence', 'wisdom', 'charisma'
    ];
    static rollAbility = () => [...Array(4).keys()].map(_i => Math.floor(Math.random() * 6 + 1))
        .sort().slice(1).reduce((acc, n) => acc + n);

    constructor() {
        this.characteristics.forEach(c => this[c] = Character.rollAbility());
        this.hitpoints = 10 + abilityModifier(this.constitution);
    }
}

File: lua/dnd-character/dnd-character.lua
local function ability(scores)
  table.sort(scores)
  return scores[4] + scores[3] + scores[2]
end
local function roll_dice()
  return { math.random(6), math.random(6), math.random(6), math.random(6) }
end
local function modifier(input)
  return (input - 10) // 2
end
local Character = {}
function Character:new (name)
  local self = setmetatable({}, Character)
  self.name = name
  self.strength = ability(roll_dice())
  self.dexterity = ability(roll_dice())
  self.constitution = ability(roll_dice())
  self.intelligence = ability(roll_dice())
  self.wisdom = ability(roll_dice())
  self.charisma = ability(roll_dice())
  self.hitpoints = 10 + modifier(self.constitution)
  return self
end
return {
  Character = Character,
  ability = ability,
  roll_dice = roll_dice,
  modifier = modifier
}
